solutions/paritosh-sharma/week-1/problem-1.2.js

```
@@ -0,0 +1,108 @@
+                continue outerLoop;
+              }
+            }
+          }
+        }
+      }
+      throw new Error(`Invalid argument '${inputArgs[ii]}' entered. Check Type and Key`);
+    }
+  var validationResult = parsedObjValidator(parseObj);
+  return parseObj;
+}
+
+// add new arg types here
+
+addArgType({ name: 'String', condition: function(argValue) { return true; }, dataType: 'String', castFunction: function(argValue) { return argValue; } });
+addArgType({ name: 'Boolean', condition: function(argValue) { return true; }, dataType: 'Boolean', castFunction: function(argValue) { return true; } });
+addArgType({ name: 'Int', condition: function(argValue) { return !isNaN(argValue); }, dataType: 'Number', castFunction: function(argValue) { return Number(argValue); } });
+addArgType({ name: '+Int', condition: function(argValue) { return (!isNaN(argValue) && argValue * 1 >= 0); }, dataType: 'Number', castFunction: function(argValue) { return Number(argValue); } });
+
+// add new arguments here in the format argBuilder({key: 'argKey', arg: argName, type: 'argType', isRequired: true/false, position: index});
+
+argGroupBuilder({ groupId: 1, isRequired: false, argList: [ { key: 'command', arg: null, type: 'String', position: 0 } ] });
+argGroupBuilder({ groupId: 2, isRequired: false, argList: [ { key: 'subcommand', arg: null, type: 'String', position: 1 } ] });
+argGroupBuilder({ groupId: 3, isRequired: true, argList: [ { key: 'key', arg: '--key', type: '+Int', position: null } ] });
+argGroupBuilder({ groupId: 4, isRequired: false, argList: [ { key: 'name', arg: '--name', type: 'String', position: null } ] });
+argGroupBuilder({ groupId: 5, isRequired: false, argList: [ { key: 'verbrose', arg: '-v', type: 'Boolean', position: null } ] });
+argGroupBuilder({ groupId: 6, isRequired: false, argList: [ { key: 'local', arg: '--local', type: 'Boolean', position: null }, { key: 'remote', arg: '--remote', type: 'Boolean', position: null } ] });
```

Uh ... this API is rather inconvenient to use. The user has to provide these complex objects with multiple keys, and they wouldn't know if they made any mistakes until they actually run your `argParser` method, and maybe not even then. There is a "fail fast" philosophy, suggesting that you want errors to be generated ASAP.

At this point, I'd suggest looking at the API provided by https://www.npmjs.com/package/commander (or other examples) to see how simple they are to use, and compare what you have with them. After that, can you improve your API?